src/snowflake/snowpark/_internal/ast/builder.py

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2025 Snowflake Computing Inc. All rights reserved.
+#
+
+import base64
+import itertools
+import sys
+import uuid
+from collections import namedtuple, deque
+from typing import Optional, List
+
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+
+from snowflake.snowpark.version import VERSION
+
+# The current AST version number (generated by the DSL).
+CLIENT_AST_VERSION = proto.__Version__.MAX_VERSION
+
+
+SerializedBatch = namedtuple("SerializedBatch", ["request_id", "batch"])
+
+
+class AstBuilder:
+    """
+    A handler for AST statements that enables a Snowpark oject (e.g. Dataframe, Table, etc.) to carry its own AST,
+    and to generate AST requests.
+
+    The core statement types are:
+    - Bind: Creates a new variable and assigns a value to it.
+    - Eval: Evaluates a variable.
+    """
+
+    # Function used to generate request IDs. This is overridden in some tests.
+    __generate_request_id = uuid.uuid4
+
+    # Class variable for generating IDs for bind statements.
+    __id_gen = itertools.count(start=1)
+
+    # NOTE: itertools.count and its __next__ method are only thread-safe in CPython with GIL.
+    # In other implementations (Jython, PyPy), this may not be thread-safe, and once the GIL is removed from CPython, it will not be thread-safe either.
+    # Likely, it would be best to make all of the methods in this class thread-safe with threading.Rlock
+    def __init__(
+        self,
+    ) -> None:
+        """
+        Initializes a new AST batch.
+        """
+        self._bind: proto.Bind = None
+        self.__eval: proto.Eval = None
+        self.__dependencies: List[AstBuilder] = []
+        self.__required_by: List[AstBuilder] = []
+
+    @classmethod
+    def Bind(cls, symbol: Optional[str] = None) -> "AstBuilder":
+        """
+        Factory method to create an AstBuilder instance with a new bind statement.
+
+        Args:
+            symbol: An optional symbol to name the new variable in the bind statement.
+        Returns:
+            An instance of AstBuilder with a bind statement initialized.
+        """
+        b = cls()
+        b._bind = proto.Bind()
+        b._bind.uid = next(AstBuilder.__id_gen)
+        b._bind.symbol.value = symbol if isinstance(symbol, str) else ""
+        return b
+
+    def add_dependency(
+        self, dependency: "AstBuilder", add_inverse: bool = False
```

Comment:
why is `add_inverse` needed? Can we remove it?

Reply:
Not necessary unless we'd like to keep track of both parent and child relations between AST statements.
Just an easy way to avoid needing another call into an `AstBuilder` method.

Reply:
Oh, I was more referring to the interface. I actually like having the double-linked relationship (i.e., you can access children/parents flexible). Would always do it/enforce it, but remove the parameter `add_inverse` from the interface. There's no need for this to exist/be an option really imho.